File: challenges/Custom_Msgs_Srvs/src/battery_status_subscriber.cpp
#include <ros/ros.h>
#include <challenge_msgs/LedControl.h>
#include <challenge_msgs/BatteryStatus.h>

void	callback_status(const challenge_msgs::BatteryStatus &msg)
{
	ros::NodeHandle nh;

	ros::ServiceClient client = nh.serviceClient<challenge_msgs::LedControl>("/led_control");
	challenge_msgs::LedControl ledControl;

	if (msg.battery_percentage == 100)
		ledControl.request.led_config = 1;
	else if (msg.battery_percentage == 50)
		ledControl.request.led_config = 2;
	else
		ledControl.request.led_config = 3;

	if (client.call(ledControl))
		ROS_INFO("Battery is at %ld%%", msg.battery_percentage);
	else
		ROS_WARN("Service call failed");
}

int	main(int argc, char **argv)
{
	ros::init(argc, argv, "Battery_Status");
	ros::NodeHandle nh;
	ros::Subscriber sub = nh.subscribe("/battery_node", 1000, callback_status);

	ros::spin();
}

File: basics/Custom_Msgs_Srvs/script/battery_status_subscriber.py
#!/usr/bin/env python3

import rospy
from challenge_msgs.msg import BatteryStatus
from challenge_msgs.srv import LedControl

percent = 0

def percent_check():
    if percent == 100:
        return 1
    if percent == 50:
        return 2
    if percent == 0:
        return 3
    return 0

def set_led():
    rospy.wait_for_service("/set_led")
    try:
        service = rospy.ServiceProxy("/set_led", LedControl)
        state = percent_check()
        resp = service(state)
    except rospy.ServiceException as e:
        rospy.logerr(e)

def callback_status(msg):
    global percent
    percent = msg.battery_percentage
    rospy.loginfo(msg.warn_message)
    set_led()

if __name__ == "__main__":
    rospy.init_node("Battery_Status")

    sub = rospy.Subscriber("/battery_node", BatteryStatus, callback_status)
    rate = rospy.Rate(2)

    rospy.spin()
